spec/lib/kafka/avro_producer_spec.rb

```diff
@@ -26,38 +26,42 @@
   let(:invalid_payload) { { 'invalid_key' => 'value' } }
   let(:schema) do
     VCR.use_cassette('kafka/topics') do
-      response = Kafka::SchemaRegistry::Service.new.subject_version('submission_trace_form_status_change_test',
-                                                                    'latest')
+      response = Kafka::SchemaRegistry::Service.new.subject_version(
+        'submission_trace_form_status_change_test', 'latest'
+      )
 
       schema = response['schema']
       Avro::Schema.parse(schema)
     end
   end
 
   before do
-    allow(Rails).to receive(:env).and_return(ActiveSupport::StringInquirer.new('test'))
-    allow(Flipper).to receive(:enabled?).with(:kafka_producer).and_return(true)
     allow(Kafka::OauthTokenRefresher).to receive(:new).and_return(double(on_oauthbearer_token_refresh: 'token'))
   end
 
   context 'using the correct client' do
     context 'in the test environment' do
-      before do
-        allow(Rails).to receive(:env).and_return(ActiveSupport::StringInquirer.new('test'))
-        Kafka::ProducerManager.instance.send(:setup_producer) # Reinitialize the producer with the mocked environment
-        allow(avro_producer).to receive(:get_schema).and_return(schema)
-      end
-
       it 'uses the Buffered client' do
         expect(avro_producer.producer.client).to be_a(WaterDrop::Clients::Buffered)
       end
     end
 
     context 'in other environments' do
-      before do
-        allow(Rails).to receive(:env).and_return(ActiveSupport::StringInquirer.new('production'))
-        Kafka::ProducerManager.instance.send(:setup_producer) # Reinitialize the producer with the mocked environment
-        allow(avro_producer).to receive(:get_schema).and_return(schema)
+      around do |example|
+        # Reset singleton before changing environment
+        Singleton.__init__(Kafka::ProducerManager)
+
+        original_env = ENV.fetch('RAILS_ENV', nil)
+        # Temporarily change environment
+        begin
+          ENV['RAILS_ENV'] = 'production'
```

Comment:
I worry this will interfere with other tests since they're run in parallel in CI. Is that a valid concern? `Rails.env` is `'production'` in all live envs (dev, staging, sandbox and prod). 

If it helps to have the specific env, you can mock `vsp_environment` like this:

```ruby
allow(Settings).to receive(:vsp_environment).and_return('production')
```

Reply:
What about this?
```ruby
allow(Rails.env).to receive(:production?).and_return(true)
```
Do you think that would work here? I think the `around do` will only protect the tests within this test thread and not other test threads. 

That being said... This has run 8 times in CI and not failed yet, so maybe it's okay? If mocking can work in this test, I do recommend it.

Reply:
I actually simplified the tests using a similar approach. 😄